### PR TITLE
Remove `existing` property - Part 4, IRuntimeFactory + RuntimeFactoryHelper 

### LINKED ISSUE #6347: Remove `existing` state from Fluid Framework - runtime
(work in progress)

See https://github.com/microsoft/FluidFramework/issues/3429 for details.

The goal here is to:
- mark the API as deprecated
- remove the state from `ContainerRuntime`, `FluidDataStoreRuntime`, `FluidDataStoreContext` and refactor as appropriate.

### PR DESCRIPTION
**Pull request by andre4i**
See #3429

- optional argument for `IRuntimeFactory.instantiateRuntime` to replace the `existing` state + an optional API compatibility marker (which will be used by the loader to decide which API to call in the backwards compatibility scenario [here](https://github.com/microsoft/FluidFramework/blob/main/packages/loader/container-loader/src/containerContext.ts#L295) as shown [here](https://github.com/microsoft/FluidFramework/pull/6486/files#diff-df051cdf51942431833eacd9aa861824a55866ee682c5880c747986a521e90c6R296)).
- `RuntimeFactoryHelper` is to be used by current clients of IRuntimeFactory (see https://github.com/microsoft/FluidFramework/pull/6486 for examples) in a future PR, after the version bump in container definitions. This class now holds the wrapper logic about how the runtime is actually instantiated.

### CODE REVIEW

File: common/lib/container-definitions/api-report/container-definitions.api.md [hunk 1]
```diff
@@ -448,7 +448,7 @@ export const IRuntimeFactory: keyof IProvideRuntimeFactory;
 
 // @public
 export interface IRuntimeFactory extends IProvideRuntimeFactory {
-    instantiateRuntime(context: IContainerContext): Promise<IRuntime>;
+    instantiateRuntime(context: IContainerContext, existing?: boolean): Promise<IRuntime>;
 }
```
Comment (andre4i) [inline common/lib/container-definitions/api-report/container-definitions.api.md:452]: The `stateful` property has been removed as it's not actually necessary. The `existing` flag is optional so a new loader can load an old runtime (with no parameter in the function definition) and just pass the parameter which will end up being ignored. So considering we're not changing the function name, there are no back-compat worries here.

File: common/lib/container-definitions/src/runtime.ts [hunk 2]
```diff
@@ -166,6 +166,9 @@ export interface IRuntimeFactory extends IProvideRuntimeFactory {
     /**
      * Instantiates a new IRuntime for the given IContainerContext to proxy to
      * This is the main entry point to the Container's business logic
+     *
+     * @param context - container context to be supplied to the runtime
+     * @param existing - whether to instantiate for the first time or from an existing context
      */
-    instantiateRuntime(context: IContainerContext): Promise<IRuntime>;
+    instantiateRuntime(context: IContainerContext, existing?: boolean): Promise<IRuntime>;
 }
```
Comment (andre4i) [inline common/lib/container-definitions/src/runtime.ts:174]: The stateful property has been removed as it's not actually necessary. The existing flag is optional so a new loader can load an old runtime (with no parameter in the function definition) and just pass the parameter which will end up being ignored. So considering we're not changing the function name, there are no back-compat worries here.

File: packages/runtime/runtime-utils/src/runtimeFactoryHelper.ts [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+    IRuntimeFactory,
+} from "@fluidframework/container-definitions";
+
+export abstract class RuntimeFactoryHelper implements IRuntimeFactory {
+    public get IRuntimeFactory() { return this; }
+
+    public async instantiateRuntime(context: IContainerContext, existing?: boolean): Promise<IRuntime> {
+        const fromExisting = existing === undefined
+            ? context.existing === true
+            : existing === true;
```
Comment (jatgarg) [inline packages/runtime/runtime-utils/src/runtimeFactoryHelper.ts:18]: You can just pass existing here. No need to check existing === true.

File: common/lib/container-definitions/src/runtime.ts (hunk 2 above)
Comment (vladsud) [reply common/lib/container-definitions/src/runtime.ts:174]: I assume that's a staging change, as you would need to make it non-optional pretty soon, right?
You could avoid this PR (and next bump to consume it) by simply casting in loader and start to pass boolean here even though TSC would not be happy (with assumption that you would follow with definition change to add this boolean and eventually remove cast in loader). It's almost same number of PRs to raise (less by one) :)

File: packages/runtime/runtime-utils/src/runtimeFactoryHelper.ts [hunk 4]
```diff
@@ -0,0 +1,35 @@
...
+            await this.instantiateFromExisting(runtime);
+        } else {
+            await this.instantiateFirstTime(runtime);
+        }
+
+        await this.hasInitialized(runtime);
+        return runtime;
+    }
+
+    public abstract preInitialize(context: IContainerContext, existing: boolean): Promise<IRuntime>;
```
Comment (vladsud) [inline packages/runtime/runtime-utils/src/runtimeFactoryHelper.ts:31]: IRuntime is not very useful interface to work with, as it looks upstream (to the loader), vs. methods like instantiateFromExisting wants to work with downstream interface (IContainerRuntime).

So we can make it return IRuntime & IContainerRuntime, which would more useful (and maybe drop IRuntime when calling methods on this class with resulting object)

We can also make it work with ContainerRuntime (concrete object), either directly or through making this class being generic (i.e. RuntimeFactoryHelper<T extends IContainerRuntime>). That might be good think if we want to get rid of IContainerRuntime (there is very little value in this interface, as code either works at this layer and knows it's ContainerRuntime, or code works with IContainerRuntimeBase - smaller version of this interface).
I'd probably defer doing this work until / if we make a choice to remove IContainerRuntime - using IContainerRuntime  for now should be sufficient.

**Comment by vladsud** [comment]
@anthony-murphy - please glance RE where to make changes to definitions.
I think we need to do them in 0.39, though I think it's not ideal - I'd rather see us bumping versions (as long as changes are back compatibly) of definition packages and changes being maid in main, then people needing to figure out (and occasionally - be wrong) on where to make them.

**Review by vladsud — APPROVED** [review]

File: common/lib/container-definitions/src/runtime.ts (hunk 2 above)
Comment (andre4i) [reply common/lib/container-definitions/src/runtime.ts:174]: Yes, that is the plan. However, I have identified some implementations of this interface in bohemia which will break if I make the change non-backwards compatible from the start.

**Pull request merged**